pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -286,89 +288,137 @@ func (n *NetworkPolicyController) processClusterNetworkPolicy(cnp *crdv1alpha1.C
+				if len(acnp.Spec.AppliedTo) > 0 {
 					// Create a rule for each affected Namespace of appliedTo at spec level
 					for i := range clusterAppliedToAffectedNS {
-						klog.V(4).Infof("Adding a new per-namespace rule with appliedTo %v for rule %d of %s", clusterAppliedToAffectedNS[i], idx, cnp.Name)
-						addRule(n.toNamespacedPeerForCRD(perNSPeers, clusterAppliedToAffectedNS[i]), direction, []string{atgForNamespace[i]})
+						klog.V(4).InfoS("Adding a new per-namespace rule with appliedTo a ClusterNetworkPolicy rule", "appliedToAffectedNS", clusterAppliedToAffectedNS[i], "ruleIndex", idx, "clusterNetworkPolicy", acnp.Name)
+						addRule(n.toNamespacedPeerForCRD(perNSPeers, clusterAppliedToAffectedNS[i]), services, direction, []string{atgForNamespace[i]})
 					}
 				} else {
 					// Create a rule for each affected Namespace of appliedTo at rule level
-					for _, at := range cnpRule.AppliedTo {
+					for _, at := range acnpRule.AppliedTo {
 						affectedNS, selectors := n.getAffectedNamespacesForAppliedTo(at)
 						affectedNamespaceSelectors = append(affectedNamespaceSelectors, selectors...)
 						for _, ns := range affectedNS {
 							atg := n.createAppliedToGroup(ns, at.PodSelector, nil, at.ExternalEntitySelector)
 							atgNamesSet.Insert(atg)
-							klog.V(4).Infof("Adding a new per-namespace rule with appliedTo %v for rule %d of %s", atg, idx, cnp.Name)
-							addRule(n.toNamespacedPeerForCRD(perNSPeers, ns), direction, []string{atg})
+							klog.V(4).InfoS("Adding a new per-namespace rule with appliedTo for a ClusterNetworkPolicy rule", "appliedToGroup", atg, "ruleIndex", idx, "clusterNetworkPolicy", acnp.Name)
+							addRule(n.toNamespacedPeerForCRD(perNSPeers, ns), services, direction, []string{atg})
 						}
 					}
 				}
 			}
+			klog.InfoS("Checking cluster or perNSPeers", "numberOfClusterPeers", len(clusterPeers), "numberOfPerNamespacePeers", len(perNSPeers))
+		} else {
+			// Handle toService rules.
+			ruleAppliedTos := acnpRule.AppliedTo
+			if appliedToPerRule && len(acnp.Spec.AppliedTo) > 0 {
+				ruleAppliedTos = acnp.Spec.AppliedTo
+			}
+			ruleATGNames := n.processClusterAppliedTo(ruleAppliedTos, atgNamesSet)
+			// Because each ServiceReference refers to its own combination of ports and
+			// Pods/IPBlocks，we need to add an individual rule for each ServiceReference.
+			for _, eachService := range acnpRule.ToServices {
+				svc, err := n.serviceLister.Services(eachService.Namespace).Get(eachService.Name)
+				if err != nil {
+					klog.V(2).InfoS("Service referred in `toServices` doesn't exist", "serviceNamespace", eachService.Namespace, "serviceName", eachService.Name)
+					continue
+				}
+				if svc.Spec.Type != v1.ServiceTypeNodePort {
+					klog.V(2).InfoS("Processing NodePort Service: still install rules matching the Endpoints IP+Ports, but no rule matching NodePort will be installed", "serviceNamespace", eachService.Namespace, "serviceName", eachService.Name)
```

Comment:
that doesn't seem right, did you mean `==` instead of `!=`. What about Services of type load balancer?

Reply:
My bad. Changed.